engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,574 @@
+  }
+
+  @Specialization
+  boolean equalsUnresolvedConversion(UnresolvedConversion selfConversion, UnresolvedConversion otherConversion,
+      @Cached EqualsAnyNode equalsNode) {
+    return equalsNode.execute(selfConversion.getScope(), otherConversion.getScope());
+  }
+
+  /**
+   * If one of the objects has warnings attached, just treat it as an objects without any
+   * warnings.
+   */
+  @Specialization(guards = {
+      "selfWarnLib.hasWarnings(selfWithWarnings) || otherWarnLib.hasWarnings(otherWithWarnings)"
+  }, limit = "3")
+  boolean equalsWithWarnings(Object selfWithWarnings, Object otherWithWarnings,
+      @CachedLibrary("selfWithWarnings") WarningsLibrary selfWarnLib,
+      @CachedLibrary("otherWithWarnings") WarningsLibrary otherWarnLib,
+      @Cached EqualsAnyNode equalsNode
+  ) {
+    try {
+      Object self =
+          selfWarnLib.hasWarnings(selfWithWarnings) ? selfWarnLib.removeWarnings(selfWithWarnings)
+              : selfWithWarnings;
+      Object other =
+          otherWarnLib.hasWarnings(otherWithWarnings) ? otherWarnLib.removeWarnings(otherWithWarnings)
+              : otherWithWarnings;
+      return equalsNode.execute(self, other);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /** Interop libraries **/
+
+  @Specialization(guards = {
+      "selfInterop.isNull(selfNull)",
+      "otherInterop.isNull(otherNull)"
+  }, limit = "3")
+  boolean equalsNull(
+      Object selfNull, Object otherNull,
+      @CachedLibrary("selfNull") InteropLibrary selfInterop,
+      @CachedLibrary("otherNull") InteropLibrary otherInterop
+  ) {
+    return true;
+  }
+
+  @Specialization(guards = {
+      "selfInterop.isBoolean(selfBoolean)",
+      "otherInterop.isBoolean(otherBoolean)"
+  }, limit = "3")
+  boolean equalsBooleanInterop(
+      Object selfBoolean,
+      Object otherBoolean,
+      @CachedLibrary("selfBoolean") InteropLibrary selfInterop,
+      @CachedLibrary("otherBoolean") InteropLibrary otherInterop
+  ) {
+    try {
+      return selfInterop.asBoolean(selfBoolean) == otherInterop.asBoolean(otherBoolean);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "selfInterop.isMetaObject(selfMeta)",
+      "otherInterop.isMetaObject(otherMeta)",
+  }, limit = "3")
+  boolean equalsMetaObjects(Object selfMeta, Object otherMeta,
+      @CachedLibrary("selfMeta") InteropLibrary selfInterop,
+      @CachedLibrary("otherMeta") InteropLibrary otherInterop,
+      @Cached EqualsAnyNode equalsNode) {
+    try {
+      Object selfMetaName = selfInterop.getMetaQualifiedName(selfMeta);
+      Object otherMetaName = otherInterop.getMetaQualifiedName(otherMeta);
+      return equalsNode.execute(selfMetaName, otherMetaName);
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
```

Comment:
Is it safe to say that two MetaObjects are same if their qualified names are the same?

It seems to me that it is quite possible to have two metaobjects with the same name, representing different stuff. Especially looking at the note:

> Sample interpretations: The qualified name of a Java class includes the package name and its class name. JavaScript does not have the notion of qualified name and therefore returns the simple name instead.

So it seems even more likely to get such conflicts in JS. Shouldn't we instead delegate to the `Object.equals` of these meta objects or to `interop.isSameObject`?

Reply:
Something like `interop.isSameObjects(meta1, meta2)` would definitely be more precise. I will try to add some tests for this particular use case. I wanted to add some `ValuesGenerator` tests anyway. I doubt that this can be reached from stdlib tests.

Reply:
Looking at the docs of meta objects in `InteropLibrary`, it seems that `interop.getMetaQualifiedName` and `interop.getMetaSimpleName` are the only methods that deal with meta objects, and are usable in this context, that we can use. We cannot use `interop.isMetaInstance` here, because we do not have the original object, for which the meta instance was fetched.

Reply:
Why would you want to use `interop.isMetaInstance` in the first place?

What about `isSameObject`?

Reply:
There is no `InteropLibrary.isSameObject` method. The only method that resembles that is `InteropLibrary.isIdentical`, which returns true iff the two objects refer to the identical instances, which is not what we want here.

If you are not comfortable with just checking qualified meta names, the only workaround I see now is to remove this `equalsMetaObjects` specialization, and let meta objects fall through to `equalsGeneric` fallback.

Reply:
Will `equalsGeneric` be correct for them? If yes - why even have the specialization in the first place?

At first glance it feels indeed safer to just fall back to the `equals` method, assuming that the Meta Object will implement it correctly. Comparing just qualified names is a simplification that adds extra assumptions (that two meta objects are equal iff their names are equal) which I'm not sure if are safe to add.

Reply:
> Will equalsGeneric be correct for them?

Yes, it will. Originally, I wanted to keep this specialization so that we will be free to provide a better (performance-wise) implementation once this specialization starts failing. If meta objects fall through to `equalsGeneric`, we might forget to optimize this case.

But I think that you are right. Let's make this semantically correct and remove this specialization. I don't think we will ever need to optimize equality for meta objects anyway.